### DISCUSSION TITLE
Extensive garbage collection when using an `event` with a connection from script

### DISCUSSION THREAD

**Discussion by DoctorGester** [ClearFoundry/ClearScript#568] [Q&A] [answered]
Hi,

We have a game server instance running in JS which is managed from C# code.

Knowing how expensive interop between C# and JS is, I have reduced the interaction scope to a minimum, however I can't figure out if it's possible to reduce the GC pressure and amount of allocations produced by C# code when invoking an event handler.

Specifically, the code looks something like this:

```cs
public delegate void OnGameTick(double time);

    public event OnGameTick? Tick;

    public void Start() {
        gameScript.Start();

        var remainingTimeLimitMs = 0L;
        var time = Stopwatch.StartNew();

        while (true) {
            var tickTimeMs = 33L;

            while (remainingTimeLimitMs >= tickTimeMs) {
                Tick?.Invoke(gameTime);
                gameTime += (tickTimeMs / 1000d);
                remainingTimeLimitMs -= tickTimeMs;
            }

            Thread.Sleep((int) tickTimeMs);

            remainingTimeLimitMs += time.ElapsedMilliseconds;
            time.Restart();
        }
    }
```

And the script trivially connects to the `Tick` handler with 

```js
    GameScript.Tick.connect(time => {
        GameRules.SetGameTime(time);
        update();
    });
```

Now, the call to `Tick?.Invoke(gameTime)` produces an intense number of allocations in the Small Object Heap, multiple gigabytes per minute. Is there a way to invoke script functions and pass data in a more garbage collector friendly way?

**Comment by ClearScriptLib** [comment]
Hi @DoctorGester,

>Now, the call to `Tick?.Invoke(gameTime)` produces an intense number of allocations in the Small Object Heap, multiple gigabytes per minute.

Hmm, can you tell what all those objects are? Also, do you know approximately how many events your application fires per minute?

Thanks!

**Reply by DoctorGester** [reply]
![image](https://github.com/microsoft/ClearScript/assets/3684394/ad21dc4f-57a8-4894-842e-697a4e200664)
![image](https://github.com/microsoft/ClearScript/assets/3684394/5a2cbafc-dd67-431c-9948-b1232cfe5fec)

@ClearScriptLib at a minimum the application is calling tick 30 times per second, so 1800 times per minute

I can share the snapshots if you wish

**Reply by ClearScriptLib** [reply]
Hi @DoctorGester,

Your screenshots suggest that there's more going on than just the firing of events into script code. Specifically, your event handler seems to call back into the host, at least sometimes. It's difficult to offer recommendations without details.

Generally speaking, .NET applications do a lot of heap allocation. Core facilities such as lambdas, LINQ, and even `foreach`-style enumeration all construct objects on the heap. And because it deals with script engines and other dynamic execution facilities, ClearScript makes heavy use of things like boxed values and parameter arrays, all of which also involve heap allocation.

To be honest, historically we've focused on preventing memory leaks and reducing working set size rather than minimizing heap allocation rate. There are things we could do to improve that – and we'll take a closer look – but there are no quick fixes.

One quick question: You said that you've minimized host-script interaction, but you're still looking at 30 hops per second plus however often the event handler calls back to the host. Would it make sense to move at least your inner loop into the script engine?

One other thing: Can you describe the effect of frequent GC on your application? Are you seeing freezes, increased latency, etc.? Much work has been done in recent runtimes to reduce GC impact on modern machines.

Thank you!

**Reply by DoctorGester** [reply]
@ClearScriptLib 

> Specifically, your event handler seems to call back into the host, at least sometimes. It's difficult to offer recommendations without details.

Right, this definitely happens, but not very often, i.e. by far not on every call of `Tick`. This also produces garbage, but since it doesn't happen often, it's not as concerning.

> ClearScript makes heavy use of things like boxed values and parameter arrays, all of which also involve heap allocation
> To be honest, historically we've focused on preventing memory leaks and reducing working set size rather than minimizing heap allocation rate

Yes, I did take a brief look into source code and I can see lots of things like delegate captures, boxing structs, throwaway IDisposable scope classes etc (it's a darn shame C# still doesn't support stack-allocated method params). I didn't expect any quickfixes to the current codepath, by the sheer number of different allocated types, but I thought maybe there is a better way to do what I'm doing? Some way to I guess "bake" my invocation of a script so it doesn't need to do so much work every time? Kinda grasping at straws here.

> One quick question: You said that you've minimized host-script interaction, but you're still looking at 30 hops per second plus however often the event handler calls back to the host. Would it make sense to move at least your inner loop into the script engine?

Is that even possible? I don't think V8 actually offers any timing primitives whatsoever by default, things like `setInterval` are the APIs provided by the user runtime i.e. in browsers and nodejs (as far as I know). 

There is a `V8.Extended` package in NuGet which offers things like `setInterval`, looking at the source code (however outdated) I can see it basically does the same thing I do here: a background thread with a wait mechanism (SpinWait instead of Thread.Sleep though) and a callback.

> One other thing: Can you describe the effect of frequent GC on your application? Are you seeing freezes, increased latency, etc.? Much work has been done in recent runtimes to reduce GC impact on modern machines.

We have not launched our game yet, but at this point we are sometimes seeing increased CPU pressure (causing latency) in a hosted server instance, so memory allocations by clearscript are one of the areas I was investigating.

**Reply by ClearScriptLib** [reply]
>Right, this definitely happens, but not very often, i.e. by far not on every call of `Tick`. This also produces garbage, but since it doesn't happen often, it's not as concerning.

OK, so the biggest issue here is the allocation impact of firing an event into the script engine. Ironically, in terms of allocation rate, firing such an event is one of ClearScript's most efficient script invocation paths, but as you've observed, its impact is still fairly steep when the event rate is in the dozens per second.

> I thought maybe there is a better way to do what I'm doing? Some way to I guess "bake" my invocation of a script so it doesn't need to do so much work every time?

Sadly, nothing viable comes to mind.

> > Would it make sense to move at least your inner loop into the script engine?
>
>Is that even possible? I don't think V8 actually offers any timing primitives whatsoever by default

That's right, but your inner loop above (the one that fires the event) doesn't use timing primitives. On the other hand, if that inner `while` isn't expected to loop many times per outer `while`, then moving it into the script engine isn't likely to be a big win.

What if you moved _both_ loops into the script engine? You mentioned timing primitives, but ClearScript could provide a synchronous delay implemented entirely in unmanaged code. What other host services would you need to invoke with similar frequency?

**Reply by DoctorGester** [reply]
> That's right, but your inner loop above (the one that fires the event) doesn't use timing primitives. On the other hand, if that inner while isn't expected to loop many times per outer while, then moving it into the script engine isn't likely to be a big win.

Yes the inner loop is not expected to be fired many times, it maybe expected to loop twice at most.

> ClearScript could provide a synchronous delay implemented entirely in unmanaged code

That would actually be great.

> What other host services would you need to invoke with similar frequency?

Not sure I understood the question here, what is meant by host services here?

**Reply by ClearScriptLib** [reply]
> > ClearScript could provide a synchronous delay implemented entirely in unmanaged code

> That would actually be great.

Great! Would a simple `sleep(milliseconds)` be sufficient? 

> > What other host services would you need to invoke with similar frequency?
> 
> Not sure I understood the question here, what is meant by host services here?

Sorry, "host services" simply refers to facilities not covered by the JavaScript environment and not amenable to a pure JavaScript implementation. A `sleep` function is one example. We were wondering if there's anything else we could provide that would enable you to move your outer loop into the script engine.

**Reply by DoctorGester** [reply]
> Great! Would a simple sleep(milliseconds) be sufficient?

Yes!

> Sorry, "host services" simply refers to facilities not covered by the JavaScript environment and not amenable to a pure JavaScript implementation

Oh I see, sure. I would also need some way to actually measure time more or less accurately, such as `performance.now()` in browsers/nodejs. I'm currently using a C# implementation of that, but that itself can incur serious overhead, so I'm not sure if it would work for ms-precision measurements.

**Reply by ClearScriptLib** [reply, marked as answer]
Hi @DoctorGester,

> > Great! Would a simple sleep(milliseconds) be sufficient?
>
> Yes!

We've looked at this and found that the even the best OS sleep primitives are a bit unpredictable, at least on Windows, so the function we're adding looks like this:

``` JavaScript
Performance.sleep(delay, precise);
```

If `precise` is `false` or unspecified, `sleep` uses an OS sleep primitive. If `precise` is `true`, it performs a "cooperative spin wait", which provides excellent precision at the cost of mild CPU load (≤3% in our tests).

Additionally, we're adding a way to increase native timer resolution, enhancing the precision of the OS sleep primitive. It's only effective on Windows, but Linux and macOS don't seem to need it anyway.

The `Performance` object will also provide a `now` method and a `timeOrigin` property that mirror similar Web and Node.js APIs. Hopefully that'll cover your timing needs. If not, please let us know!

Cheers!

**Reply by DoctorGester** [reply]
Great, that sounds like exactly what we need!

**Reply by ClearScriptLib** [reply]
[Version 7.4.5](https://github.com/microsoft/ClearScript/releases/tag/7.4.5) added `AddPerformanceObject` and `SetTimerResolution` to [`V8ScriptEngineFlags`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_V8_V8ScriptEngineFlags.htm).
